stl/inc/ranges

```diff
@@ -6837,6 +6837,602 @@ namespace ranges {
         inline constexpr _Stride_fn stride;
     } // namespace views
 
+#ifdef __cpp_lib_ranges_zip
+    namespace _Zip {
+        template <typename... _RangeTypes>
+        concept _Is_common = (sizeof...(_RangeTypes) == 1 && (common_range<_RangeTypes> && ...)) //
+                          || (!(bidirectional_range<_RangeTypes> && ...) && (common_range<_RangeTypes> && ...)) //
+                          || ((random_access_range<_RangeTypes> && ...) && (sized_range<_RangeTypes> && ...));
+
+        template <typename _CallbackType>
+        constexpr auto _Create_tuple_transform_lambda(_CallbackType&& _Callback) noexcept {
+            const auto _Created_lambda = [&_Callback]<typename... _ViewTupleTypes>(_ViewTupleTypes&&... _View_tuples) {
+                return tuple<invoke_result_t<_CallbackType&, _ViewTupleTypes>...>{
+                    _STD invoke(_Callback, _STD forward<_ViewTupleTypes>(_View_tuples))...};
+            };
+
+            return _Created_lambda;
+        }
+
+        template <typename _CallbackType, typename _TupleType>
+        constexpr auto _Tuple_transform(_CallbackType&& _Callback, _TupleType&& _Tuple) noexcept(
+            noexcept(_STD apply(_Zip::_Create_tuple_transform_lambda(_STD forward<_CallbackType>(_Callback)),
+                _STD forward<_TupleType>(_Tuple)))) /* strengthened */ {
+            return _STD apply(_Zip::_Create_tuple_transform_lambda(_STD forward<_CallbackType>(_Callback)),
+                _STD forward<_TupleType>(_Tuple));
+        }
+
+        template <typename _CallbackType>
+        constexpr auto _Create_tuple_for_each_lambda(_CallbackType&& _Callback) noexcept {
+            const auto _Created_lambda = [&_Callback]<typename... _ViewTupleTypes>(_ViewTupleTypes&&... _View_tuples) {
+                (_STD invoke(_Callback, _STD forward<_ViewTupleTypes>(_View_tuples)), ...);
+            };
+
+            return _Created_lambda;
+        }
+
+        template <typename _CallbackType, typename _TupleType>
+        constexpr void _Tuple_for_each(_CallbackType&& _Callback, _TupleType&& _Tuple) noexcept(
+            noexcept(_STD apply(_Zip::_Create_tuple_for_each_lambda(_STD forward<_CallbackType>(_Callback)),
+                _STD forward<_TupleType>(_Tuple)))) /* strengthened */ {
+            _STD apply(_Zip::_Create_tuple_for_each_lambda(_STD forward<_CallbackType>(_Callback)),
+                _STD forward<_TupleType>(_Tuple));
+        }
+
+        template <size_t _DesiredOffset, size_t _FirstIndex, size_t... _OtherIndices>
+        struct _Offset_index_sequence_solver {
+            static constexpr auto _Solved_index_sequence{
+                _Offset_index_sequence_solver<_DesiredOffset, _OtherIndices...>::_Solved_index_sequence};

```

Comment:
Although this is correct, we generally don't introduce `static constexpr` data members when computing tag types like this. Instead, we say `using type = STUFF;`. We construct an object of that tag type only at the top-level function call where we need it.

(Note that `type` and `value` are the two major exceptions where we use non-`_Ugly` names in internal machinery, as there's no way they can be macroized without breaking the universe, and SFINAE mistakes caused by confusing `type` and `_Type` would be annoying.)

Reply:
Alright, that makes sense. Thanks for pointing out those exceptions to the identifier naming practices! Should a description for this be added to the `PR Checklist` in `CONTRIBUTING.md`? I suppose such a scenario is too situational to warrant the extra text, but it appears to be common enough for me to have made this mistake.

Reply:
I'll note this in the Contribution Guidelines that I need to write up when I get a chance. Thanks!